### PR TITLE
Swap out spray for Akka http

### PR DESCRIPTION
**Pull request by seanpquig**
Apologies, didn't mean to edit your description - Russell

### CODE REVIEW

**Comment by rcoh** [comment]
Thanks for this long needed change! It's not building on CI for some reason -- I assume it builds for you locally?

**Comment by seanpquig** [comment]
It does build, and all tests run for me locally @rcoh.  However, I'm using Java 8.  I can revert to Akka versions that still support Java 7?  Unless there are plans to migrate to Java 8 for the project?

File: elasticsearch-aws/src/main/scala/com/sumologic/elasticsearch/util/AwsRequestSigner.scala [hunk 1]
```diff
@@ -171,7 +179,8 @@ class AwsRequestSigner(awsCredentialsProvider: AWSCredentialsProvider, region: S
   }
 
   private def hashedPayloadByteArray(httpRequest: HttpRequest): Array[Byte] = {
-    val data = httpRequest.entity.asString
+    val dataFuture = AkkaHttpUtil.entityToString(httpRequest.entity)
+    val data = Await.result(dataFuture, 1.second)
```
Comment (rcoh) [inline elasticsearch-aws/src/main/scala/com/sumologic/elasticsearch/util/AwsRequestSigner.scala]: what precisely is being waited for here? Are we waiting for the request to finish or is it just for some sort of string conversion?

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala [hunk 2]
```diff
@@ -225,16 +226,18 @@ class RestlasticSearchClient(endpointProvider: EndpointProvider, signer: Option[
...
-    val responseFuture: Future[HttpResponse] = (IO(Http) ? request).mapTo[HttpResponse]
+    val responseFuture: Future[HttpResponse] = Http().singleRequest(request)
 
-    responseFuture.map { response =>
+    responseFuture.flatMap { response =>
       logger.debug(f"Got Es response: $response")
+      val entityFuture = AkkaHttpUtil.entityToString(response.entity)
       if (response.status.isFailure) {
-        logger.warn(s"Failure response: ${response.entity.asString.take(500)}")
+        val entityString = Await.result(entityFuture, 100.millis)
```
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala]: Maybe cleaner to use `.recover` or `.recoverWith` to handle the failure without having to Await on the entity?

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/util/AkkaHttpUtil.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    * Helper method for getting string out of Akka HTTP Entity
+    * @param entity HTTP entity
+    * @param streamTimeout timeout for Strict stream completion
+    * @param ec implicit ExecutionContext
+    */
+  def entityToString(entity: ResponseEntity,
+                     streamTimeout: FiniteDuration = 1000.millis)
+                    (implicit ec: ExecutionContext = ExecutionContext.global,
+                     materializer: ActorMaterializer): Future[String] = {
+    entity.toStrict(streamTimeout).map(_.data.utf8String)
```
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/util/AkkaHttpUtil.scala]: I would prefer doing the `Await` here -- we wait for seemingly arbitrary amounts of time for this in other places. I'd rather standardize in one place even if it makes the API a little less flexible. Does this only return a future as a semantic side effect of akka-http or is there really something that we're waiting for?

**Review by rcoh — CHANGES_REQUESTED** [review]

**Comment by CCheSumo** [comment]
@rcoh Regarding the java version, I have seen two PRs having problem with java 7. As a matter of fact, our infrastructure is also moving to java 8. What do you think about we upgrade the scala/akka version? I will create a PR if that sounds reasonable.

File: elasticsearch-aws/src/main/scala/com/sumologic/elasticsearch/util/AwsRequestSigner.scala (hunk 1 above)
Comment (seanpquig) [reply elasticsearch-aws/src/main/scala/com/sumologic/elasticsearch/util/AwsRequestSigner.scala]: This is an artifact of all Akka HTTP request/response entities being modeled as streams, so in this instance the full payload string is being pulled from that stream abstraction.  It does not involve an HTTP request/network call.

More on streaming nature of entities: http://doc.akka.io/docs/akka-http/current/scala/http/implications-of-streaming-http-entity.html#implications-of-streaming-http-entities

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala (hunk 2 above)
Comment (seanpquig) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala]: I'll try that 👍

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/util/AkkaHttpUtil.scala (hunk 3 above)
Comment (seanpquig) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/util/AkkaHttpUtil.scala]: This future is a side effect of Akka HTTP modeling entities as streams.  I'll try refactor to better consolidate necessary waiting, and think through all the implications.

**Comment by rcoh** [comment]
@CCheSumo upgrading the scala/java version is fine with me

**Review by rcoh — APPROVED** [review]

**Comment by seanpquig** [comment]
How do you guys want to coordinate Java 8 switch with Akka versions in this PR?  Java 8 would enable these changes to use stable Akka HTTP and not the old experimental jar.

**Comment by CCheSumo** [comment]
@seanpquig let's keep the changes in this PR. After Java 8 is enabled, we have serval libs we will upgrade, the akka core included.

**Comment by seanpquig** [comment]
Sounds good @CCheSumo.  Thanks.

**Pull request merged**